libs/math/include/math/clustering/knn.hpp

```diff
@@ -27,36 +27,61 @@ namespace details {
 
 template <typename ArrayType,
           typename ArrayType::Type (*Distance)(ArrayType const &, ArrayType const &)>
-std::vector<std::pair<typename ArrayType::SizeType, typename ArrayType::Type>> GetKNNImplementation(
-    ArrayType array, ArrayType one_vector, typename ArrayType::SizeType k)
+std::vector<std::pair<typename ArrayType::SizeType, typename ArrayType::Type>> KNNImplementation(
+    ArrayType array, ArrayType vec, typename ArrayType::SizeType k)
 {
   using DataType = typename ArrayType::Type;
   using SizeType = typename ArrayType::SizeType;
 
-  assert(one_vector.shape().size() == 2);
+  assert(vec.shape().size() == 2);
   assert(array.shape().size() == 2);
-  assert(array.shape().at(1) == one_vector.shape().at(1));
-  assert(one_vector.shape().at(0) == 1);
+
+  // test vector must be either {1, N} or {N, 1}
+  assert((vec.shape().at(0) == 1) || (vec.shape().at(1) == 1));
+
+  // array must be either {M, N} (if vector shape is {1, N}) or {N, M}
+  assert(((array.shape().at(1) == vec.shape().at(1)) && (vec.shape().at(0) == 1)) ||
+         ((array.shape().at(0) == vec.shape().at(0)) && (vec.shape().at(1) == 1)));
+
+  // vec <1, M>; array <N, M>
+  //  assert(array.shape().at(1) == vec.shape().at(1));
+
+  SizeType feature_axis;
+  SizeType data_axis;
+  if (vec.shape().at(0) == 1)
+  {
+    feature_axis = 1;
+  }
+  else
+  {
+    feature_axis = 0;
+  }
+  data_axis = 1 - feature_axis;
 
   std::vector<std::pair<SizeType, DataType>> similarities;
-  similarities.reserve(array.shape().at(0));
-  for (SizeType i(0); i < array.shape().at(0); ++i)
+  std::vector<std::pair<SizeType, DataType>> ret;
+
+  similarities.reserve(array.shape().at(data_axis));
+
+  // compute distances
+  for (SizeType i = 0; i < array.shape().at(data_axis); ++i)
   {
-    typename ArrayType::Type d = Distance(one_vector, array.Slice(i).Copy());
+    DataType d = Distance(vec, array.Slice(i, data_axis).Copy());
```

Comment:
It's not possible now, but we should get rid of Copy and replace .Slice with .View, but View will need to take offset in order to work same way as Slice. Maybe we will not even need Slice at all.
DataType d=Distance(vec, array.View(data_axis,i));

Reply:
suggested change will be part of an upcoming PR as discussed